File: Java-Implementation/LinkedList/LinkedList.java
package Linked;
/**
*
*
*
**/

import java.util.*;
public class LinkedList<T>{
	//head,tail,size,element implementation

	private class Node<T>{
		private Node<T> next;
		private T element;

		protected Node(T elem, Node<T> n){
			element = elem;
			next = n;
		}

		protected Node<T> getNext(){
			return next;
		}

		protected void setNext(Node<T> n){
			next = n;
		}

		/**
		* Returns the current nodes element
		**/
		protected T getElement(){
			return element;
		}
	}

	private Node<T> head= null;
	private Node<T> tail= null;
	private int size=0;

	public LinkedList(){
		//do nothing
	}

	public int getSize(){
		return size;
	}

	public void increment(){
		size++;
	}

	public void decrement(){
		if(size>0)
			size--;
		else
			return;
	}

	public boolean isEmpty(){
		return (size==0);
	}

	public void pushFront(T elem){
		head = new Node<>(elem,head);
		if(size==0){
			tail = head; 
		}
		increment();
	}

	public void pushBack(T elem){
		Node<T> temp = new Node<>(elem,null);

		if(size==0){
			head = temp;
		}else{
			tail.setNext(temp);
		}

		tail = temp;
		increment();
	}

	public void pushIndex(T elem, int i){
		if(size!=0 && size>i && i>=0){
			Node<T> temp = head;
			Node<T> temp2;
			for(int x=0;x<i-1;x++){
				temp = temp.getNext();
			}
			temp2 = temp.getNext();
			temp.setNext(new Node<>(elem,temp2));
			increment();
		}else{
			//out of bounds
			throw new IndexOutOfBoundsException();
		}
	}

	public T popFront(){
		if(size>0){
			T temp = head.getElement();
			head = head.getNext();
			decrement();
			return temp;
		}else{
			return null;
		}
	}

	public T popBack(){
		if(size>0){
			Node<T> temp = head;
			T temp2 = tail.getElement();
			for(int x=0;x<size-1;x++){
				temp = temp.getNext();
			}
				tail = temp;
				decrement();
				return temp2;
		}else{
			return null;
		}
	}

	public void printElements(){
		Node<T> temp = head;
		for(int x=0;x<size;x++){
			System.out.print(temp.getElement() + " ");
			temp = temp.getNext();
		}
		System.out.println("");
	}

	public T elemAt(int index){
		if(index<size && index>=0){
			Node<T> temp = head;
			for(int x=0;x<index;x++){
				temp = temp.getNext();
			}
			return temp.getElement();
		}else{
			return null;
		}
	}

	public boolean searchList(T elem){
		Node<T> temp = head;
		for(int x=0;x<size;x++){
			if(temp.getElement()==elem)
				return true;
			temp = temp.getNext();
		}
		return false;
	}


}

File: C++-Implementation/LinkedList/linkedlist.c
#include "linkedlist.h"

linkedlist::linkedlist(){
	Node* head = nullptr;
	Node* tail = nullptr;
	int size = 0;
}


linkedlist::~linkedlist(){
	for(int x=0;x<size;x++){
		removeLast();
	}

	head = nullptr;
	tail = nullptr;
}

linkedlist::Node::Node(){
	Node* next = nullptr;
	int elem = 0;
}

linkedlist::Node::~Node(){
	printf("%s\n", "Deconstructing node");
}


void linkedlist::insertLast(int elem){
	Node* to_insert = new Node;
	to_insert->elem = elem;
	size++;
	if(head==nullptr){
		head = to_insert;
		tail = to_insert;
	}else{
		Node* temp = tail;
		temp->next = to_insert;
		tail = to_insert;
	}
}


void linkedlist::insertAt(int index, int elem){
	Node* to_insert = new Node;
	to_insert->elem = elem;
	if(!(index>size || index<0)){ //index is inbounds
		int curr = 0;
		Node* current = head;
		Node* temp;
		while(curr!=index){
			current = current->next;
		}
		temp = current->next;
		current->next = to_insert;
		to_insert->temp;
		if(size-1==index){
			tail = to_insert;
		}
	}else if(index==0){
		to_insert->next = head;
		head = to_insert;
	}else{ //not in bounds
		printf("%s\n", "Not in bounds!");
	}
}


int linkedlist::removeLast(){
	Node* current = head;
	while(current!=head){
		current = current->next;
	}
	int the_return = current->next->elem;
	delete current->next;
	tail = current;
	return the_return;
}


int linkedlist::removeAt(int index){
	if(size-1==index){
		return removeLast();
	}else if(!(index>size || index<0)){ //index is inbounds
		int curr = 0;
		Node* current = head;
		Node* temp;
		while(curr!=index){
			current = current->next;
		}
		temp = current->next;
		current->next = temp->next;
		delete temp;
	}else if(index==0){
		Node temp = head->next;
		delete head;
		head = temp;
	}else{ //not in bounds
		printf("%s\n", "Not in bounds!");
	}
}


int linkedlist::getSize(){
	return size;
}


bool linkedlist::findElement(int elem){
	Node current = head;
	for(int x=0;x<size;x++){
		if(current->elem == elem){
			return true;
		}else{
			current = current->next;
		}
	}

	return false; //will be called if for loop is left
}





